### PR TITLE
Refactor: use `Literals` in type hints

### PR DESCRIPTION
**Pull request by srini047**
Fixes: #1670 

- I have made sure to `protocol` with `Literals` and options being ['pickle', 'protobuf']

### CODE REVIEW

**Comment by JoanFM** [comment]
We need you to comply with our contributing guidelines and signoff the commits.

https://github.com/docarray/docarray/blob/main/CONTRIBUTING.md

**Comment by JoanFM** [comment]
> @JoanFM I have signed the commits. Could you have a look at it?
> 
> image

THE CI is complaining about that, as well as mypy, black and ruff

**Comment by srini047** [comment]
> > @JoanFM I have signed the commits. Could you have a look at it?
> > image
> 
> THE CI is complaining about that, as well as mypy, black and ruff

Well how do I proceed with. Any leads appreciated.
Thanks in advance!

File: docarray/store/helpers.py [hunk 1]
```diff
@@ -131,12 +132,27 @@ def close(self):
 
 
 def _to_binary_stream(
+    self, protocol: Literal['pickle', 'protobuf'],
```
Comment (samsja) [inline docarray/store/helpers.py:135]: default value should still be `protobuf` here and we should not change the order of the argument (protocol should stay at the same place as before)

File: docarray/base_doc/mixins/io.py [hunk 2]
```diff
@@ -146,7 +146,7 @@ def __bytes__(self) -> bytes:
         return self.to_bytes()
 
     def to_bytes(
-        self, protocol: str = 'protobuf', compress: Optional[str] = None
+        self, protocol: Literal['pickle', 'protobuf'], compress: Optional[str] = None
```
Comment (samsja) [inline docarray/base_doc/mixins/io.py:149]: we need to keep the default vlaue

**Review by samsja — CHANGES_REQUESTED** [review]
Thanks for you contribution ! Indeed litteral makes more sense in most of the case.

Two overall comments:
* Lets keep default value
* If possible we should not change the order of the parameters of existing function

File: docarray/base_doc/mixins/io.py (hunk 2 above)
Comment (srini047) [reply docarray/base_doc/mixins/io.py]: Updated the changes on all the occurrences.

File: docarray/store/helpers.py (hunk 1 above)
Comment (srini047) [reply docarray/store/helpers.py:135]: I would want to make it clear if the `protocol` is in the right place and order (i.e. 2nd argument).

Is my understanding correct? At the same time let me make sure to add `protobuf ` as the default value.

Comment (srini047) [reply docarray/store/helpers.py:135]: Also, if we want `protocol` to act as a default argument then it has to be placed at the end.

Comment (samsja) [reply docarray/store/helpers.py:135]: I think you can do what you did on the other and kept it at the end

File: docarray/store/helpers.py [hunk 3]
```diff
@@ -167,13 +183,28 @@ def _to_binary_stream(
 
 
 def _from_binary_stream(
+    self, protocol: Literal['pickle', 'protobuf'],
```
Comment (samsja) [inline docarray/store/helpers.py:186]: same here we can do like you id in the other

**Review by samsja — COMMENTED** [review]
It starts to look very good. 

One idea, maybe we could wrap the `Literal['protobuf','pickle']` into a more formal type that we could call each time instead or repeating it ?

Something like this:

```python

Protocols = Literal['pickle','protobuf']

def my_func(protocol: Protocol):
   ...
```

Wdyt ?

**Comment by srini047** [comment]
> It starts to look very good.
> 
> One idea, maybe we could wrap the `Literal['protobuf','pickle']` into a more formal type that we could call each time instead or repeating it ?
> 
> Something like this:
> 
> ```python
> Protocols = Literal['pickle','protobuf']
> 
> def my_func(protocol: Protocol):
>    ...
> ```
> 
> Wdyt ?

It increases the modularity of the code and maybe easier to maintain as you wish, what do you say?

File: docarray/store/helpers.py (hunk 3 above)
Comment (srini047) [reply docarray/store/helpers.py:186]: image

Facing this error When I try to commit the changes. Any help on the same appreciated.

**Comment by samsja** [comment]
> > It starts to look very good.
> > One idea, maybe we could wrap the `Literal['protobuf','pickle']` into a more formal type that we could call each time instead or repeating it ?
> > Something like this:
> > ```python
> > Protocols = Literal['pickle','protobuf']
> > 
> > def my_func(protocol: Protocol):
> >    ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Wdyt ?
> 
> It increases the modularity of the code and may be easier to maintain as well, what do you say?

then lets do it

File: docarray/store/helpers.py (hunk 3 above)
Comment (srini047) [reply docarray/store/helpers.py:186]: @samsja Can you have a look at this? Meanwhile let me work on the function implementation.

**Comment by srini047** [comment]
Hey @samsja!
- Please have a look at the approach if this helps. Or else let me revert to any other logic
- Also, I need help in `docarray/store/helpers.py` and I have tagged you in the issue I am facing.

File: docarray/base_doc/mixins/io.py [hunk 4]
```diff
@@ -146,7 +147,10 @@ def __bytes__(self) -> bytes:
         return self.to_bytes()
 
     def to_bytes(
-        self, protocol: str = 'protobuf', compress: Optional[str] = None
+        self,
+        data,
```
[the remark this answers is not in the document]
Comment (samsja) [reply docarray/base_doc/mixins/io.py:151]: what is this data ?

[the remark this answers is not in the document]
Comment (srini047) [reply docarray/base_doc/mixins/io.py:151]: `data` is base64 encoded string. This is the same parameter available as in `from_base64` function.
I am actually using it to `_decompress_bytes`

[the remark this answers is not in the document]
Comment (srini047) [reply docarray/base_doc/mixins/io.py:151]: @samsja Any updates?
Just tagging to ensure this doesn't remain the backlog.

**Comment by samsja** [comment]
@srini047 can you install the pre commit hook and apply black, ruff and mypy for the CI to pass !

**Pull request closed without merging**